# Schedule for Differential expression analysis of single-cell RNA-seq workshop

## Pre-reading



## Day 1

| Time |  Topic  | Instructor |
|:-----------:|:----------:|:--------:|
| 09:30 - 09:45 | [Workshop introduction](../slides/workshop_intro_slides.pdf) | Meeta |
| 09:45 - 10:35| [Introduction to Single Cell RNA-sequencing: a practical guide](../slides/072624_Chan_scRNAseq_workshop_Arpita_Kulkarni.pdf) | [Dr. Arpita Kulkarni](https://singlecellcore.hms.harvard.edu/people/arpita-kulkarni-phd) |
| 10:35 - 10:40 | Break |
| 10:40 - 11:00 | scRNA-seq pre-reading discussion | All |
| 11:00 - 11:45 | [Quality control set-up](../lessons/03_SC_quality_control-setup.md) | Noor |
| 11:45 - 12:00 | Overview of self-learning materials and homework submission | Meeta |


### Before the next class:

I. Please **study the contents** and **work through all the code** within the following lessons:
 
   1. [Quality control of cellranger counts](../lessons/04_cellranger_QC.md)
      <details>
       <summary><i>Click here for a preview of this lesson</i></summary>
       <br>Before you start any analysis, itâ€™s important to know whether or not you have good quality cells. At these early stages you can flag or remove samples that could produce erroneous results downstream. <br><br>In this lesson you will:<br>
             - Discuss the outputs of cellranger and how to run it <br>
             - Review web summary HTML report<br>
             - Create plots from metrics_summary.csv file <br><br>
        </details>

 2.  [Quality control with additional metrics](../lessons/04_SC_quality_control.md)
      <details>
       <summary><i>Click here for a preview of this lesson</i></summary>
         <br>In addition to the QC generated by cellranger, we can also compute some of our own metrics based on the raw data we have loaded into our Seurat object. <br><br>In this lesson you will:<br>
             - Compute essential QC metrics for each sample<br>
             - Create plots to visualize metrics across cells per sample<br>
             - Critically evaluate each plot and learn what each QC metric means<br><br>
        </details>
     
   3. [Theory of PCA](../lessons/05_theory_of_PCA.md)
      <details>
       <summary><i>Click here for a preview of this lesson</i></summary>
         <br>Before we can begin the next steps of the workflow, we need to make sure you have a good understanding of Principal Components Analysis (PCA). This method will be utilized in the scRNA-seq analysis workflow, and this foundation will help you better navigate those steps and interpretation of results.<br><br>
        </details>
         

II. **Submit your work**:
   * Each lesson above contains exercises; please go through each of them.
   * **Submit your answers** to the exercises using [this Google form](https://docs.google.com/forms/d/e/1FAIpQLSf4zw37eUcuNCmHxCLikg3JENoxGWIuv5c_5pqHdYNId81Arg/viewform) on **the day *before* the next class**.
   


### Questions?
* ***If you get stuck due to an error*** while runnning code in the lesson, [email us](mailto:hbctraining@hsph.harvard.edu) 

***

## Day 2

| Time |  Topic  | Instructor |
|:-----------:|:----------:|:--------:|
| 09:30 - 10:15 | Self-learning lessons discussion | All |
| 10:15 - 11:15|  [Normalization and regressing out unwanted variation](../lessons/06_SC_SCT_normalization.md) | Noor |
| 11:15 - 11:25 | Break |
| 11:25 - 12:00| [A brief introduction to Integration](../lessons/06a_integration_cca_theory.md) | Meeta |

### Before the next class:
I. Please **study the contents** and **work through all the code** within the following lessons:

1. [Running CCA integration and complex integration tasks](../lessons/06b_integration_code_harmony.md)
      <details>
       <summary><i>Click here for a preview of this lesson</i></summary>
        <br>In class, we described the theory of integration and in what situations we would implement it. <br><br>In this lesson you will:<br>
             - Run the code to implement CCA integration <br>
             - Evaluate the effect of integration on the UMAP <br>
             - Learn about methods for complex integration tasks (Harmonizing samples) <br>
        </details>
   
2. [Clustering](../lessons/07_SC_clustering_cells_SCT.md)
      <details>
       <summary><i>Click here for a preview of this lesson</i></summary>
         <br>From the UMAP visualization of our data  we can see that the cells are positioned into groups. Our next task is to isolate clusters of cells that are most similar to one another based on gene expression. <br><br>In this lesson you will:<br>
             - Learn the theory behind clustering and how it is performed in Seurat<br>
             - Cluster cells and visualize them on the UMAP<br>
        </details>

3. [Clustering quality control](../lessons/08_SC_clustering_quality_control.md)
      <details>
       <summary><i>Click here for a preview of this lesson</i></summary>
         <br>After separating cells into clusters, it is crtical to evaluate whether they are biologically meaningful or not. At this point we can also decide if we need to re-cluster and/or potentialy go back to a previous QC step.
         <br><br>In this lesson you will:<br>
           - Check to see that clusters are not influenced by uninteresting sources of variation<br>
           - Check to see whether the major principal components are driving the different clusters<br>
           - Explore the cell type identities by looking at the expression for known markers across the clusters.<br>
        </details>


II. **Submit your work**:
   * Each lesson above contains exercises; please go through each of them.
   * **Submit your answers** to the exercises using [this Google form](https://docs.google.com/forms/d/e/1FAIpQLSd8-k5YwfQekQQdFuDVPhGCy-eK70CqPiUkDNYDeWSEQYBAZg/viewform?usp=sf_link) on **the day *before* the next class**.

### Questions?
* ***If you get stuck due to an error*** while runnning code in the lesson, [email us](mailto:hbctraining@hsph.harvard.edu) 


***


## Day 3

| Time |  Topic  | Instructor |
|:-----------:|:----------:|:--------:|
| 9:30 - 10:00 | Self-learning lessons discussion | All |
| 10:00 - 11:00 |[Marker identification](../lessons/09_merged_SC_marker_identification.md) | Noor |
| 11:00 - 11:10 | Break |
| 11:10 - 11:30 | [Workflow summary](../lessons/scRNAseq_workflow.md) | Meeta |
| 11:30 - 11:45 | [Seurat Cheatsheet](../lessons/seurat_cheatsheet.md) Overview and Final Q & A | All |
| 11:45- 12:00 | [Wrap up](../slides/Workshop_wrapup.pdf) | Meeta |

***

## Answer Keys





## Resources
